tools/cloud_storage_manifest_patcher/main.py

```diff
@@ -309,46 +370,103 @@ def list_buckets(self) -> dict[str, Union[list, dict]]:
             raise
 
 
+class ManifestIterator:
+    def __init__(self, obj):
+        self.data = obj
+        self.namespace = obj['namespace']
+        self.topic = obj['topic']
+        self.partition = obj['partition']
+        self.revision = obj['revision']
+        segments = []
+        for key, value in obj['segments'].items():
+            segments.append(
+                (key, value.get('archiver_term'), value.get('ntp_revision')))
+        self.segments = sorted(segments, key=lambda x: int(x[0].split('-')[0]))
+
+    def list_segments(self):
+        for key, term, rev in self.segments:
+            yield key, get_segment_path(self.namespace, self.topic,
+                                        self.partition, rev or self.revision,
+                                        key, term)
+
+
+def get_segment_meta(s3client: S3Client, bucket, path):
+    print(f"Downloading segment {path} from {bucket}")
+    stream = s3client.get_object_data_stream(bucket, path)
+    # 'Header', ('header_crc', 'batch_size', 'base_offset', 'type', 'crc',
+    #            'attrs', 'delta', 'first_ts', 'max_ts', 'producer_id',
+    #            'producer_epoch', 'base_seq', 'record_count'))
+    rdr = SegmentReader(stream)
+    total_size = 0
+    num_configs = 0
+    base_offset = None
+    base_timestamp = None
+    max_timestamp = -1
+    last_offset = 0
+    for batch in rdr:
+        if batch is None:
+            break
+        if batch.type != 1:
+            num_configs += 1
+        max_timestamp = max(max_timestamp, batch.max_ts)
+        base_timestamp = base_timestamp or batch.first_ts
+        base_offset = base_offset or batch.base_offset
+        total_size += batch.batch_size
+        last_offset = max(last_offset, batch.base_offset + batch.record_count)
+
+    return base_offset, base_timestamp, last_offset, max_timestamp, num_configs, total_size
```

Comment:
Seems like a decent case for `typing.NamedTuple`. (This can be done as followup PR if it helps.)